core/iwasm/interpreter/wasm_loader.c

```diff
@@ -3610,6 +3610,17 @@ load_function_section(const uint8 *buf, const uint8 *buf_end,
 #endif
             }
 
+            /* Code size in code entry can't be smaller than size of vec(locals)
+             * + expr(at least 1 for opcode end). And expressions are encoded by
+             * their instruction sequence terminated with an explicit 0x0B
+             * opcode for end. */
+            if (p_code_end - 1 < p_code || *(p_code_end - 1) != WASM_OP_END) {
+                set_error_buf(
+                    error_buf, error_buf_size,
+                    "section size mismatch: function body END opcode expected");
+                return false;
+            }
+
```

Comment:
For me, it is better check `WASM_OP_END` when learning `code_size`. It will be L3572

Reply:
Whether the code ends with `WASM_OP_END` will be checked in the end of wasm_loader_prepare_bytecode, seems here we don't need to check again?
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_loader.c#L15839-L15850

And is it better to use `p_code_end <= p_code` instead of `p_code_end - 1 < p_code`?

And had better also add similar check (bh_assert) in wasm_mini_loader.c.

Reply:
My suggestion is not right. Here is how wasm-interp in wabt execute:

``` shell
$ /opt/wabt-1.0.36/bin/wasm-interp -v ./clusterfuzz-testcase-minimized-wasm_mutator_fuzz_loader-5465932459212800 
BeginModule(version: 1)
  BeginTypeSection(6)
    OnTypeCount(1)
    OnFuncType(index: 0, params: [f32], results: [i32])
  EndTypeSection
  BeginFunctionSection(2)
    OnFunctionCount(1)
    OnFunction(index: 0, sig_index: 0)
  EndFunctionSection
  BeginCodeSection(24)                       // not check END here as I guessed
    OnFunctionBodyCount(1)        
    BeginFunctionBody(0, size:2)
    OnLocalDeclCount(1)                      // check local here
    OnLocalDecl(index: 0, count: 32, type: f64)   
                                             // Expected the END for function body before 
                                             // reading next?
000001b: error: function body must end with END opcode 
```

Reply:
You mean that wabt checks whether the last byte of the code is `END` first? It may be good, but it is a little strange that normally the loader checks the bytecode one by one, if the last byte is not `END`, it may be an invalid bytecode (e.g. a unknown opcode or a malformed byte of a bytecode), so it is hard to say what we should report. wabt's behavior may be not standard compatible, could we build out the interpreter of wasm spec and use it to test?

Reply:
> You mean that wabt checks whether the last byte of the code is END first

No. I used to think so. But the verbose information shows another way.

Reply:
>  could we build out the interpreter of wasm spec and use it to test?
The spec interpreter output differs from wabt. 
```bash
./wasm -t -d ../../wasm-micro-runtime/clusterfuzz.wasm
-- Running ("(input \"../../wasm-micro-runtime/clusterfuzz.wasm\")")...
-- Parsing...
-- Running...
-- Loading (../../wasm-micro-runtime/clusterfuzz.wasm)...
-- Decoding...
-- Running...
-- Decoding...
-- Error: 
../../wasm-micro-runtime/clusterfuzz.wasm:0x2e: decoding error: unexpected end of section or function
```

Reply:
OK, if the spec interpreter also checks whether the code ends with op END first, this PR is good.